Changes in module 'nf-core/eggnogmapper'
--- modules/nf-core/eggnogmapper/meta.yml
+++ modules/nf-core/eggnogmapper/meta.yml
@@ -60,3 +60,6 @@
       pattern: "versions.yml"
 authors:
   - "@vagkaratzas"
+maintainers:
+  - "@vagkaratzas"
+  - "@gallvp"

--- modules/nf-core/eggnogmapper/main.nf
+++ modules/nf-core/eggnogmapper/main.nf
@@ -1,6 +1,6 @@
 process EGGNOGMAPPER {
     tag "$meta.id"
-    label 'process_long'
+    label 'process_high'
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
@@ -23,11 +23,13 @@
     task.ext.when == null || task.ext.when
 
     script:
-    def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
-    def is_compressed = fasta.name.endsWith(".gz")
-    def fasta_name = fasta.name.replace(".gz", "")
-    def dbmem = task.memory.toMega() > 40000 ? '--dbmem' : ''
+    def args            = task.ext.args                 ?: ''
+    def prefix          = task.ext.prefix               ?: "${meta.id}"
+    def is_compressed   = fasta.name.endsWith(".gz")
+    def fasta_name      = fasta.name.replace(".gz", "")
+    def dbmem           = task.memory.toMega() > 40000  ? '--dbmem'                         : ''
+    def database_arg    = eggnog_db                     ? "--database $eggnog_db"           : ''
+    def dmnd_db_arg     = eggnog_diamond_db             ? "--dmnd_db $eggnog_diamond_db"    : ''
     """
     if [ "$is_compressed" == "true" ]; then
         gzip -c -d $fasta > $fasta_name
@@ -38,8 +40,8 @@
         -i ${fasta_name} \\
         --data_dir ${eggnog_data_dir} \\
         -m diamond \\
-        --dmnd_db ${eggnog_diamond_db} \\
-        --database ${eggnog_db} \\
+        $dmnd_db_arg \\
+        $database_arg \\
         --output ${prefix} \\
         ${dbmem} \\
         $args

************************************************************
